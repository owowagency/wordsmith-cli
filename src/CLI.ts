import {Repository} from "./Repository";
import {Config} from "./Config";
import {Storage} from "./Storage";
import {StorageTranslationCollection} from "./StorageTranslationCollection";
import fs from "fs";
import {TranslationCollection} from "./TranslationCollection";

/**
 * A class that contains all functionality of the CLI.
 */
export class CLI {

    /**
     * Reads translations from remote repository.
     */
    static async pull(repository: Repository, config: Config): Promise<void> {
        if (! config.isPathValid()) {
            try {
                fs.mkdirSync(config.get('translations_directory'), { recursive: true })
            } catch (error) {
                throw new Error("Invalid translations directory: " + config.get('translations_directory'));
            }
        }

        let rawProjectTranslations: Object[] = await repository.getRawProjectTranslations();

        let translationCollection: TranslationCollection = new TranslationCollection(rawProjectTranslations);

        let storage: Storage = new Storage(config.get('translations_directory'));

        storage.writeTranslationsForAllLanguages(translationCollection);
    }

    /**
     * Stores translations in remote repository.
     */
    static async push(repository: Repository, config: Config): Promise<void> {
        if (! config.isPathValid()) {
            throw new Error("Invalid translations directory: " + config.get('translations_directory'));
        }

        let storage: Storage = new Storage(config.get('translations_directory'));

        let storageTranslationCollection: StorageTranslationCollection = storage.readTranslations();

        let translationsJson: any = storageTranslationCollection.toJson();

        let success: boolean = await repository.sendRawProjectTranslations(translationsJson);

        if (! success) {
            throw new Error("Error sending translations to the server.");
        }
    }
}