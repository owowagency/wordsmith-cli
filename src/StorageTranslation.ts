export class StorageTranslation {

    private readonly _language: string;

    private readonly translationValues: StorageTranslationValue[];

    constructor(language: string, translations: Record<string, string>) {
        this._language = language;
        this.translationValues = Object.keys(translations).map((key: string) => {
            return new StorageTranslationValue(key, translations[key]);
        });
    }

    public keys(): string[] {
        return this.translationValues.map((translationValue: StorageTranslationValue) => translationValue.key);
    }

    public getTranslationValue(key: string): string | undefined {
        let translationValue =
            this.translationValues.find((translationValue: StorageTranslationValue) => translationValue.key === key);
        return translationValue?.value;
    }

    get language(): string {
        return this._language;
    }
}

class StorageTranslationValue {
    key: string;
    value: string;

    constructor(key: string, value: string) {
        this.key = key;
        this.value = value;
    }
}