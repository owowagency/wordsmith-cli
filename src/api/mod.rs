use log::{error, log_enabled, debug};
use reqwest::{ClientBuilder, header::{HeaderMap, HeaderValue, InvalidHeaderValue}, Client, StatusCode, Request, Response};
use serde::de::DeserializeOwned;
use thiserror::Error;

use crate::environment::AccessToken;

use self::models::error::ApiError;

pub mod models;
pub mod pull;
pub mod push;
pub mod info;

const BASE_URL: &'static str = env!("BASE_URL");
const USER_AGENT: &'static str = env!("USER_AGENT");

pub struct WordsmithClient {
    headers: HeaderMap,
    client: Client,
}

type Result<T> = std::result::Result<T, WordsmithError>;

#[derive(Debug, Error)]
pub enum WordsmithError {
    #[error("Unable to create API Client, reason: {0}")]
    Init(String),
    #[error("HTTP Client error, reason: {0}")]
    Http(String),
    #[error("API Error, status: {0}, message: {1}")]
    Api(StatusCode, ApiError),
    #[error("IO Error, reason: {0}")]
    Io(String),
}

impl From<InvalidHeaderValue> for WordsmithError {
    fn from(value: InvalidHeaderValue) -> Self {
        WordsmithError::Init(value.to_string())
    }
}

impl From<reqwest::Error> for WordsmithError {
    fn from(value: reqwest::Error) -> Self {
        let status = value.status();
        let url = value.url();
        let message = format!("HTTP error [status: {:?}, url: {:?}], message: {}", status, url, value.to_string());
        WordsmithError::Http(message)
    }
}

impl WordsmithClient {
    pub fn new(token: Option<&AccessToken>) -> Result<Self> {
        let user_agent = HeaderValue::from_str(USER_AGENT)?;
        let accept = HeaderValue::from_str("application/json")?;
        let mut headers = HeaderMap::new();
        headers.append("User-Agent", user_agent);
        headers.append("Accept", accept);
        if let Some(token) = token {
            let raw_token = token.get_token().map_err(|op| WordsmithError::Init(op))?;
            let token_header = HeaderValue::from_str(&format!("Bearer {}", raw_token))?;
            headers.append("Authorization", token_header);
        }
        let client = ClientBuilder::new()
            .default_headers(headers.clone())
            .build()
            .map_err(|op| WordsmithError::Init(op.to_string()))?;
        
        Ok(Self {
            client,
            headers,
        })
    }

    async fn send(&self, request: Request) -> Result<Response> {
        if log_enabled!(log::Level::Debug) {
            let url = request.url();
            let query = url.query_pairs();
            let headers = request.headers();
            let mut buffer = String::new();
            buffer.push_str(&format!("{} {}", request.method(), request.url()));
            for (name, value) in self.headers.iter() {
                buffer.push_str(&format!("\n{}: {}", name.as_str(), value.to_str().ok().unwrap_or("")))
            }
            for (name, value) in headers.iter() {
                buffer.push_str(&format!("\n{}: {}", name.as_str(), value.to_str().ok().unwrap_or("")))
            }
            for (key, value) in query.into_iter() {
                buffer.push_str(&format!("\n{}={}", key, value))
            }
            debug!("{}", buffer);
        }
        return Ok(self.client.execute(request).await?);
    }

    pub async fn execute<Res : DeserializeOwned>(&self, request: Request) -> Result<Res> {
        let response = self.send(request).await?;

        let status = response.status();

        match status.as_u16() {
            200..=299 => {
                let response: Res = response.json().await?;
                Ok(response)
            },
            _ => Err(WordsmithClient::decode_error_response(response).await),
        }
    }

    pub async fn execute_no_content(&self, request: Request) -> Result<()> {
        let response = self.send(request).await?;
        let status = response.status();

        if status != StatusCode::NO_CONTENT {
            return Err(WordsmithClient::decode_error_response(response).await);
        }

        Ok(())
    }

    pub async fn download(&self, url: &str) -> Result<Vec<u8>> {
        let request = self.client.get(url).build()?;
        let response = self.send(request).await?;
        let status = response.status();

        if status != StatusCode::OK {
            return Err(WordsmithClient::decode_error_response(response).await);
        }

        return Ok(response.bytes().await?.to_vec());
    }

    async fn decode_error_response(response: Response) -> WordsmithError {
        let status = response.status();
        let content_type = match response.headers().get("Content-Type") {
            Some(header) => header.to_str().unwrap_or(""),
            None => "",
        };

        if content_type.contains("application/json") {
            return match response.json().await {
                Ok(error) => WordsmithError::Api(status, error),
                Err(err) => WordsmithError::Api(status, ApiError { 
                    message: err.to_string(),
                    errors: vec![],
                })
            };
        }

        match response.text().await {
            Ok(text) => WordsmithError::Api(status, ApiError {
                message: text,
                errors: vec![],
            }),
            Err(err) => err.into(),
        } 
    }
}

