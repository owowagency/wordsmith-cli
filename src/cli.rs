use clap::{command, Parser, Subcommand, Args};

use crate::environment::Environment;

#[derive(Parser)]
#[command(author, version, about, long_about = None)]
pub struct CommandLine {
    #[command(subcommand)]
    pub command: Command,
}

#[derive(Subcommand)] 
pub enum Command {
    Pull(PullArgs),
    Push(PushArgs),
}

#[derive(Debug, Args)]
pub struct GlobalArgs {
    #[arg(short, long, default_value = "wordsmith.yml", value_parser = clap::value_parser!(Environment))]
    pub env: Environment,
    /// Enable verbose logging
    #[arg(long, default_value_t = false)]
    pub verbose: bool
}

#[derive(Debug, Args)]
pub struct PushArgs {
    #[clap(flatten)]
    pub global: GlobalArgs,
    /// Overwrites existing strings
    #[arg(short = 'f', long = "force", default_value_t = false)]
    pub overwrite: bool,
    /// Verify translations
    #[arg(short, long, default_value_t = false)]
    pub verify: bool,
    #[arg(short = 'd', long = "dry-run", default_value_t = false)]
    pub dry_run: bool,
}

#[derive(Debug, Args)]
pub struct PullArgs {
    #[clap(flatten)]
    pub global: GlobalArgs,
    #[arg(short = 'd', long = "dry-run", default_value_t = false)]
    pub dry_run: bool,
}