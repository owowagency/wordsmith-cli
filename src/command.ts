export const test = () => {
  console.log('yoooo');
}
