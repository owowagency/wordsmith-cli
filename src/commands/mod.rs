use async_trait::async_trait;

use crate::{cli::Command, api::WordsmithError};

mod list_projects;
mod pull;
mod push;
mod helpers;

#[async_trait]
pub trait Execute {
    async fn execute(&self) -> Result<(), WordsmithError>;
}

#[async_trait]
impl Execute for Command {
    async fn execute(&self) -> Result<(), WordsmithError> {
        match self {
            Command::ListProjects(args) => args.execute().await?,
            Command::Pull(args) => args.execute().await?,
            Command::Push(args) => args.execute().await?,
        };

        Ok(())
    }
}
