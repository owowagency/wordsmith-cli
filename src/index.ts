const { program } = require("commander");
const packageJson = require('../package.json');

import { Config } from './config';

let config = new Config();

program
    .name(config.get('cli.name'))
    .description(packageJson.description)
    .version(packageJson.version)
    .option('-token, --api-token <token>', 'Set the token used to authenticate with the API')
    .option('-id, --project-id <id>', 'Set the ID of the project that will be used for translations')
    .option('-d, --translations-directory <directory>', 'Set the directory where the translations will be stored')
    .option('-t --tags <tags>', 'Set the tags that will be used to filter the translations');

program.parse(process.argv);

const options = program.opts();

if (options.apiToken) { config.set('client.api_token', options.apiToken); }

if (options.projectId) { config.set('client.project_id', options.projectId); }

if (options.translationsDirectory) { config.set('client.translations_directory', options.translationsDirectory); }

if (options.tags) { config.set('client.tags', options.tags); }

if (! config.isConfigured()) {
    console.log("Please configure the CLI before using it.");
    program.help();
}