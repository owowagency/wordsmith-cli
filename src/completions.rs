use std::fs::File;
use std::io::Write;
use std::path::Path;

mod cli;
mod environment;
mod commands;
mod api;

use api::WordsmithError;
use clap::{CommandFactory, Args, Parser};
use clap_complete::{Generator, generate, shells};
use cli::CommandLine;
use log::error;

#[derive(Parser)]
#[command(author, version, about, long_about = None)]
struct GenerateCli {
    #[arg(short, long)]
    shell: Variant,
    #[arg(short, long)]
    output: String,
}


impl GenerateCli {
    fn execute(&self) -> Result<(), WordsmithError> {
        let path = Path::new(&self.output);
        let mut output = File::create(path)?;

        match self.shell {
            Variant::Bash => generate_completion(shells::Bash, &mut output),
            Variant::Fish => generate_completion(shells::Fish, &mut output),
            Variant::Zsh => generate_completion(shells::Zsh, &mut output),
            Variant::Powershell => generate_completion(shells::PowerShell, &mut output),
        };

        Ok(())
    }
}

pub fn generate_completion<G: Generator>(shell: G, buf: &mut dyn Write) {
    let mut command = CommandLine::command();
    generate(
        shell,
        &mut command,
        env!("CARGO_BIN_NAME"),
        buf,
    );
}

#[derive(Debug, Args)]
struct GenerateCompletionsArgs {
    #[arg(short, long)]
    pub shell: Variant,
    #[arg(short, long)]
    pub output: String,
}

#[derive(Clone, Debug, clap::ValueEnum)]
enum Variant {
    Bash,
    Fish,
    Zsh,
    Powershell,
}

fn main() {
    let app = GenerateCli::parse();
    match app.execute() {
        Ok(_) => {},
        Err(err) => error!("Failed to generate completions, cause: {}", err),
    };
}