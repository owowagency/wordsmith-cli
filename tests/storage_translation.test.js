import {StorageTranslation} from "../src/StorageTranslation";
import expect from "expect";

const storageLithuanianTranslationsJson = {
    "promiscuous girl": "pasileidusi mergina",
    "bladee": "Dzordana Butkute"
};

const storageDutchTranslationsJson = {
    "promiscuous girl": "promiscue meisje",
    "bladee": "Waylon"
}

const storageLithuanianTranslations = new StorageTranslation("LT", storageLithuanianTranslationsJson);

const storageDutchTranslations = new StorageTranslation("NL", storageDutchTranslationsJson);

test('constructor creates storage translation from translations json', () => {
        expect(storageLithuanianTranslations).toMatchSnapshot();

        expect(storageDutchTranslations).toMatchSnapshot();
});

test('keys returns all keys', () => {
        expect(storageLithuanianTranslations.keys()).toEqual(["promiscuous girl", "bladee"]);
});

test('get translation value returns translation value for a given key', () => {
        expect(storageLithuanianTranslations.getTranslationValue("promiscuous girl"))
            .toEqual("pasileidusi mergina");

        expect(storageDutchTranslations.getTranslationValue("promiscuous girl"))
            .toEqual("promiscue meisje");
});

