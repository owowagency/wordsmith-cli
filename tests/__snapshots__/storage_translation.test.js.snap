// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`constructor creates storage translation from translations json 1`] = `
StorageTranslation {
  "_language": "LT",
  "translationValues": [
    StorageTranslationValue {
      "key": "promiscuous girl",
      "value": "pasileidusi mergina",
    },
    StorageTranslationValue {
      "key": "bladee",
      "value": "Dzordana Butkute",
    },
  ],
}
`;

exports[`constructor creates storage translation from translations json 2`] = `
StorageTranslation {
  "_language": "NL",
  "translationValues": [
    StorageTranslationValue {
      "key": "promiscuous girl",
      "value": "promiscue meisje",
    },
    StorageTranslationValue {
      "key": "bladee",
      "value": "Waylon",
    },
  ],
}
`;
